index.js

```diff
@@ -305,8 +305,12 @@ async function wasm_pack(cx, state, dir, source, id, options) {
                     import * as exports from ${import_path};
                     ${prelude}
 
-                    export default async () => {
-                        await exports.default(${import_wasm});
+                    export default async (input) => {
+                        if (typeof input === 'function') {
```

Comment:
I think this is better implemented like this:

```js
let path = ${import_wasm};

if (input != null) {
    path = input(path);
}

await exports.default(path);
```

Reply:
I updated the code but kept support for string. I think it's valuable to support the same API that `input` exported by wasm bindgen already supports. Otherwise, the code generated by rollup will not work with e.g. the types generated by wasm bindgen.

Reply:
This plugin does not generate TypeScript types (since Rollup doesn't support that).

And the types are already different from wasm-bindgen: this plugin uses the types from `index.d.ts`, but wasm-bindgen uses the types from `index_bg.d.ts`.

And in any case wasm-bindgen is an implementation detail, this plugin has its own API and options (which already differ from wasm-bindgen's options).

Reply:
> This plugin does not generate TypeScript types (since Rollup doesn't support that).

I see. Isn't that an argument for supporting string as this way one can copy the types from wasm bindgen (I am doing that in https://github.com/domoritz/arrow-wasm/blob/6547069228d49f8bc4e261f88a1633b71ad93dbb/rollup.config.js#L16)? Yes, the types are not fully correct but at least the behavior of init generated by this rollup plugin is similar to the behavior of init from wasm bindgen. I feel that using an object is a premature optimization as we don't have anything else to pass for now. Also, we could always support an object in the future if it becomes necessary and we wouldn't have to break backward compatibility.

The final call is of course with you. Let me know and I'll leave the pull request or update it as you request.

Reply:
> Isn't that an argument for supporting string as this way one can copy the types from wasm bindgen

Unfortunately I don't know of any way to make that work, it's not as simple as copying the types, TypeScript has a completely different file resolving algorithm.

And if it were possible, then I still wouldn't copy the types, instead I would create a new `.d.ts` file. The types simply do not match between wasm-bindgen and this plugin.

> I feel that using an object is a premature optimization as we don't have anything else to pass for now.

That's not what premature optimization means. Making designs forward-extensible is very common and very sensible. It is a major mistake to lock designs in too early.

> Also, we could always support an object in the future if it becomes necessary and we wouldn't have to break backward compatibility.

I don't want to add in cruft which will be changed later, I prefer to do things right the first time. That is the Rust attitude.

> The final call is of course with you. Let me know and I'll leave the pull request or update it as you request.

I really would like it to use an object.

Reply:
>> I feel that using an object is a premature optimization as we don't have anything else to pass for now.

> That's not what premature optimization means. Making designs forward-extensible is very common and very sensible. It is a major mistake to lock designs in too early.

We could still support extra parameters but I guess the fact that js doesn't (yet) have named arguments makes it better to have an object argument. I updated the pull request with an object.